### PR TITLE
refactor: perform single traverse for state checks

### PR DESCRIPTION
**Pull request by massfords**
It's cleaner performing the state specific checks in a single traverse of the definition.

### CODE REVIEW

File: src/checks/state-checks.ts [hunk 1]
```diff
@@ -0,0 +1,98 @@
+import {getStates, getStatesContainer, StateEntry, StateFilter} from "./get-states";
+import {State, StateMachine, StateMachineError, StateMachineErrorCode, ValidationOptions} from "../types";
+
+export const IsMap = ({state}: StateEntry): boolean => {
+    return state.Type === 'Map'
+}
```
Comment (massfords) [inline src/checks/state-checks.ts:6]: helper functions to make the checks in the validate func a little more readable

File: src/checks/state-checks.ts [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+export const IsChoice = ({state}: StateEntry): boolean => {
+    return state.Type === 'Choice'
+}
+
+export type StateChecker = (entry: StateEntry) => StateMachineError | null
+
+export const stateChecks = (
+    definition: StateMachine,
+    options: ValidationOptions,
+    checks: Array<{ filter: StateFilter, checker: StateChecker }>): StateMachineError[] => {
```
Comment (massfords) [inline src/checks/state-checks.ts:33]: performs the single traversal and performs the checks on the applicable states

File: src/checks/state-checks.ts [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+}
+
+const getPropertyCount = ({props, state}: {
+    props: string[],
+    state: State,
+}) : number => {
+    return props.map((prop) => prop in state ? 1 : 0)
+        .map((val) => Number(val))
+        .reduce((prev, curr) => prev + curr, 0)
+}
```
Comment (massfords) [inline src/checks/state-checks.ts:61]: moved here from the previous checkProperty workflow

File: src/checks/state-checks.ts [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+export const AtMostOne = ({props, errorCode}:{props: string[], errorCode: StateMachineErrorCode}): StateChecker => {
+    return ({state, stateName}) => {
+        const count = getPropertyCount({state, props})
+        if (count > 1) {
+            return {
```
Comment (massfords) [inline src/checks/state-checks.ts:67]: same code as before, just moved. 

this and the helper func below `ExactlyOne` produce checkers used in the validator

File: src/validator.ts [hunk 5]
```diff
@@ -30,10 +37,54 @@ export = function validator(definition: StateMachine, opts?: ValidationOptions):
...
-        errors.push(...waitDuration(definition, options));
-        errors.push(...taskChecks(definition, options));
-        errors.push(...mapChecks(definition, options));
+        errors.push(...stateChecks(definition, options, [
+            {
+                filter: IsTask,
+                checker: AtMostOne({
+                    props: ["TimeoutSeconds", "TimeoutSecondsPath"],
+                    errorCode: StateMachineErrorCode.TaskTimeoutError})
+            },
```
Comment (massfords) [inline src/validator.ts:46]: removed the separate functions in favor of inlined logic here for the state checks

note that the `filter` and `checker` params are both `functions`. The code here only uses the helper functions `AtMostOne` and `ExactlyOne` but custom filters/checkers are possible

**Review by ChristopheBougere — APPROVED** [review]
LGTM

**Pull request merged**